### PR TITLE
Initial implementation of GRU layers

### PR DESCRIPTION
**Pull request by ptigwe**
Currently this shows a working implementation of GRU layer which follows quite closely the Keras API, using the `flax.linen.GRUCell` as its backbone.

Tackling the implementation of GRU (#40)

### CODE REVIEW

**Comment by cgarciae** [comment]
Thanks @ptigwe for this! Sorry it took so long to respond, I'd seen it but hadn't had the time. Overall it looks very good, I'll just leave a couple of comments.

File: treex/nn/recurrent.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+        Returns:
+            - The final state of the GRU by default
+            - The full sequence of states (if `return_sequences == True`)
+            - A tuple of both the sequence of states and final state (if both
+                `return_state` and `return_sequences` are `True`)
+        """
+        # Move time dimension to be the first so it can be looped over
+        if not self.time_major:
+            x = jnp.transpose(x, (1, 0, 2))
```
Comment (cgarciae) [inline treex/nn/recurrent.py:151]: `swap_axes`

File: treex/nn/recurrent.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+        if self.go_backwards:
+            x = x[::-1, :, :]
```
Comment (cgarciae) [inline treex/nn/recurrent.py:154]:
```
x = x[::-1]
```

File: treex/nn/recurrent.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        [flax_module.PRNGKey, flax_module.Shape, flax_module.Dtype], flax_module.Array
+    ]
+    bias_init: tp.Callable[
+        [flax_module.PRNGKey, flax_module.Shape, flax_module.Dtype], flax_module.Array
+    ]
+    initial_state_init: tp.Callable[
+        [flax_module.PRNGKey, flax_module.Shape, flax_module.Dtype], flax_module.Array
+    ]
+    params: tp.Dict[str, tp.Dict[str, flax_module.Array]] = types.Parameter.node()
+    last_state: flax_module.Array = types.BatchStat.node()
```
Comment (cgarciae) [inline treex/nn/recurrent.py:40]: Unsure if this should be a `BatchStat` node, these are intended to be synced during distributed training. I think `types.Cache.node()` is a better fit.

**Comment by cgarciae** [comment]
As discussed offline, we feel here we can lift Keras restrictions regarding the shape of the input by removing the `time_major` argument in favor of something like a `time_axis` which lets you select the dimension you are going to `scan` over. This changes requires a couple of changes to the logic including figuring out the proper `batch_dims` to `initialize_carry`.

**Comment by cgarciae** [comment]
Hey @ptigwe! 
There are a few minor comments / changes left if you'd like to finish them else I'd be very happy to continue with the PR.

**Comment by ptigwe** [comment]
@cgarciae, I believe I have fixed most of the comments which you mentioned before, including the `time_axis` as discussed. The only thing I didn't update was the switch from `jax.lax.scan` to `jax_utils.scan_in_dim`, as I was waiting for the change in flax to get merged and the upstream version updated.

Please let me know if there is anything else which I missed and I would make the changes ASAP.

**Comment by cgarciae** [comment]
Thanks @ptigwe ! There is a small comment about changing the default `time_axis` from `0` to `-2`.

**Comment by ptigwe** [comment]
@cgarciae, I guess I must have missed that one. I've added it now to the PR.

**Comment by cgarciae** [comment]
Keras uses `[B, T, C]`, we are going for `[..., B, T, C]` which is even better. I think this is nicer than `[...,T, B, C]` because 1D convolutions and transformers also use `[..., B, T, C]` so the op would chain better.

**Comment by ptigwe** [comment]
OK cool. It has already been set in the previous commit. Also decided not to include the change to `jax_utils.scan_in_dim` as the current version being pointed to `0.3.26` does not have the `unrolled` added to it yet.

**Comment by cgarciae** [comment]
If you want to update `flax` to the latest version to add this feature now you can run:

```bash
poetry add flax@latest
```

**Comment by ptigwe** [comment]
Seems the `latest` is indeed `0.3.6` which indeed does not have the updated `scan_in_dims`. Whenever it gets updated we can always update the code to make use of it. I've also included some comments on things that might need changing in that case.

**Comment by cgarciae** [comment]
@ptigwe Sounds good. I'll merge this for now, we can create a new PR with `scan_in_dims` latter. 

Thanks a lot for pushing this through!

**Pull request merged**